New: Add new GSS updates for the next Lab release (#883)
memgraph/docs: lab/style-script/overview.md

We can highlight nodes with low drink prices in the following way. We want to
use a beer image and a bigger size along with a red shadow.

```cpp
@NodeStyle And(
     HasLabel(node, "City"),
     Less(Property(node, "drinks_USD"), 5)) {
  size: 50
  image-url: "https://www.sciencenews.org/wp-content/uploads/2020/05/050620_mt_beer_feat-1028x579.jpg"
  shadow-color: red
 }
```

## Caching results for faster performance

> Sometimes you might want to normalize some value, for example, the size or width of all the
> nodes or relationships in the graph. To do so, you need to find the minimum and maximum values
> of all nodes.
> Let's say we have nodes labeled `"Person"` with a property `age` that holds the age information
> of a particular person. We want a node property `size` to be 5 for the youngest person
> and 20 for the oldest one in the presented graph. All other node sizes should be normalized
> within that range.
> 
> One of the solutions could look like this:

Reviewer: ```suggestion
To normalize some value, for example, the size or width of all the
nodes or relationships in the graph, find the minimum and maximum values
of all nodes.
For example, a node labled "Person"` has the property `age` that holds the age information
of a particular person. We want the node property `size` to be 5 for the youngest person
and 20 for the oldest one in the presented graph. All other node sizes should be normalized
within that range.

One of the solutions could look like this:
```

We can highlight nodes with low drink prices in the following way. We want to
use a beer image and a bigger size along with a red shadow.

```cpp
@NodeStyle And(
     HasLabel(node, "City"),
     Less(Property(node, "drinks_USD"), 5)) {
  size: 50
  image-url: "https://www.sciencenews.org/wp-content/uploads/2020/05/050620_mt_beer_feat-1028x579.jpg"
  shadow-color: red
 }
```

## Caching results for faster performance

> Sometimes you might want to normalize some value, e.g. size or width of all the
> nodes/edges in the graph. To do so, you need to find the minimum and maximum values
> of all the nodes.
> Let's say we have nodes `"Person"` with a property `age` that holds the age information
> of a particular person. We want a node property `size` to be 5 for the youngest person
> and 20 for the oldest one in the presented graph. All other node sizes should be normalized
> within that range.

Reviewer: ```suggestion
Sometimes you might want to normalize some value, for example, the size or width of all the
nodes or relationships in the graph. To do so, you need to find the minimum and maximum values
of all nodes.
Let's say we have nodes labeled `"Person"` with a property `age` that holds the age information
of a particular person. We want a node property `size` to be 5 for the youngest person
and 20 for the oldest one in the presented graph. All other node sizes should be normalized
within that range.
```

One of the solutions could look like this:

```cpp
// Size range min/max variables
Define(MIN_SIZE, 5)
Define(MAX_SIZE, 20)
Define(PROP_NAME, "age")
Define(SIZE_RANGE, Sub(MAX_SIZE, MIN_SIZE))

// A set of utility functions
// Create a new array of property values from an array of nodes
Define(GetProperties, Function(nodes, propName,
  Map(nodes, Function(singleNode, Property(singleNode, propName)))
))
// Keep only the numeric values from an array of values
Define(KeepNumericValues, Function(values,
  Filter(values, Function(value, IsNumber(value)))
))

// Functions to find min and max value in the input nodes
Define(GetMaxValue, Function(nodes,
  Max(KeepNumericValues(GetProperties(nodes, PROP_NAME)))
))
Define(GetMinValue, Function(nodes,
  Min(KeepNumericValues(GetProperties(nodes, PROP_NAME)))
))

// Normalize function that receives two inputs: node (n) and
// graph (g) and returns normalized value into a range
// [MIN_SIZE, MAX_SIZE]
Define(Normalize, Function(n, g,
  Add(
    MIN_SIZE,
    Mul(
      SIZE_RANGE,
      Div(
        Sub(Property(n, PROP_NAME), GetMinValue(Nodes(g))),
        Sub(GetMaxValue(Nodes(g)), GetMinValue(Nodes(g)))
      )
    )
  )
))

> // For all the nodes with label "Person" and numeric property "age"

Reviewer: ```suggestion
// For all nodes with the label "Person" and numeric property "age"
```

@NodeStyle And(HasLabel(node, "Person"), IsNumber(Property(node, PROP_NAME))) {
  color: white
  size: Normalize(node, graph)
  width: Div(Normalize(node, graph), 5)
  label: Format("Age: {}", AsText(Property(node, PROP_NAME)))
}
```

![Using Graph Style Script to style different nodes by its size](../data/caching-results-gss.png)

> The problem with the above solution is the performance. `Normalize` is called twice
> for each node in the graph view. Each `Normalize` call is iterating through all the
> nodes three times: two calls of `GetMinValue` and a single call `GetMaxValue`.
> For small graphs, you won't see a difference in the performance, but try running this
> on a larger set of nodes.
> 
> To solve this issue, cache the results by calculating outside of the `@NodeStyle` and
> `@EdgeStyle` directives. Variable `graph` is available outside of these directives which
> can be utilized.
> Within the `@NodeStyle` directive, local variable can be used to store the normalized
> value which is then used in `size` and `width` property. We don't call `Normalize` twice
> anymore for each node, but only once.
> 
> Check the improved GSS code below:

Reviewer: ```suggestion
The problem with the solution above is slow performance. The `Normalize` function is called twice
for each node in the graph view. Each `Normalize` call iterates through all nodes three times: two two times for `GetMinValue` and once for `GetMaxValue`.
For small graphs, you won't see a difference in performance but as the number of nodes rises the performance issues will follow. 

To solve this issue, cache the results by calculating outside of `@NodeStyle` and
`@EdgeStyle` directives where the variable `graph` is also available.
Inside the `@NodeStyle` directive, a local variable can be used to store the normalized
value and use it with `size` and `width` properties thus calling the `Normalize` function only once.

Check the improved GSS code below:
```

```cpp
// Size range min/max variables
Define(MIN_SIZE, 5)
Define(MAX_SIZE, 20)
Define(PROP_NAME, "age")
Define(SIZE_RANGE, Sub(MAX_SIZE, MIN_SIZE))

// A set of utility functions
// Create a new array of property values from an array of nodes
Define(GetProperties, Function(nodes, propName,
  Map(nodes, Function(singleNode, Property(singleNode, propName)))
))
// Keep only the numeric values from an array of values
Define(KeepNumericValues, Function(values,
  Filter(values, Function(value, IsNumber(value)))
))

// Variables MAX_VALUE and MIN_VALUE will hold the max and min

> // value from all node properties.

Reviewer: ```suggestion
// values of all node properties.
```

> // If is used to handle errors when there are no values to calculate

Reviewer: ```suggestion
// The If statement is used to handle errors when there are no values to calculate
```

> // min and max on.

Reviewer: ```suggestion
// min and max from.
```

> // min and max on.

Reviewer: ```suggestion
// min and max from.
```
